File: LC/java/Best Time to Buy and Sell Stock III.java
public class Solution {
    public int maxProfit(int[] prices) {
        int maxP;
        int[] f = new int[prices.length];
        int minV;
        int[] b = new int[prices.length];
        int maxV;
        
        if(prices.length == 0){
            return 0;
        }
        minV = prices[0];
        maxP = Integer.MIN_VALUE;
        for(int i = 0; i < prices.length; i++){
            minV = Math.min(minV,prices[i]);
            maxP = Math.max(maxP,prices[i]-minV);
            f[i] = maxP;
        }
        maxV = prices[prices.length-1];
        maxP = Integer.MIN_VALUE;
        for(int i = prices.length-1; i>-1; i--){
            maxV = Math.max(maxV,prices[i]);
            maxP = Math.max(maxP,maxV-prices[i]);
            b[i] = maxP;
        }
        maxP = f[f.length-1];
        for (int i = 0; i < prices.length-1; i++){
            maxP = Math.max(maxP, f[i]+b[i+1]);
        }
        return maxP;
    }
}

File: LC/python/Best Time to Buy and Sell Stock III.py
class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        if len(prices) == 0:
            return 0
            
        f = []
        minV = prices[0]
        maxP = -1
        for i in prices:
            minV = min(i,minV)
            maxP = max(i-minV,maxP)
            f.append(maxP)
        
        b = []
        maxV = prices[len(prices)-1]
        maxP = -1
        for i in reversed(prices):
            maxV = max(i,maxV)
            maxP = max(maxV-i,maxP)
            b.insert(0,maxP)
        
        res = f[-1] 
        for i in range (len(f)-1):
            res = max(res,f[i]+b[i+1])
        return res
            
        
